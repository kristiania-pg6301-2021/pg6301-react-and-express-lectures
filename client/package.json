{
  "name": "client",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "start": "parcel src/index.html",
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "react": "^17.0.2",
    "react-dom": "^17.0.2",
    "react-router": "^5.2.0",
    "react-router-dom": "^5.2.0"
  },
  "devDependencies": {
    "parcel": "^2.0.0-beta.2",
    "prettier": "^2.2.1"
  }
}
