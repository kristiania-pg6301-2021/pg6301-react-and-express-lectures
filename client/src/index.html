<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <title>KristianiaQuiz</title>
  </head>
  <body>
    <div id="app"></div>
  </body>
  <script src="index.jsx"></script>
</html>
