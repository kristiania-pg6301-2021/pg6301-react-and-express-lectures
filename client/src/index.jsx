import React from "react";
import ReactDOM from "react-dom";
import { BrowserRouter, Link } from "react-router-dom";
import { Route, Switch } from "react-router";

function Application() {
  return (
    <BrowserRouter>
      <header>
        <Link to={"/"}>KristianiaQuiz</Link>
      </header>
      <main>
        <Switch>
          <Route path="/login">
            <h1>Login</h1>
          </Route>
          <Route path="/" exact>
            <h1>Home page</h1>
            <ul>
              <li>
                <Link to={"/login"}>Login</Link>
              </li>
            </ul>
          </Route>
        </Switch>
      </main>
    </BrowserRouter>
  );
}

ReactDOM.render(<Application />, document.getElementById("app"));
