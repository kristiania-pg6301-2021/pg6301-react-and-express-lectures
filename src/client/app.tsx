import * as React from "react";

import {BrowserRouter} from "react-router-dom";
import {Route, Switch} from "react-router";

interface Quiz {
    question: string;
    alternatives: string[];
}


function HomePage() {
    return <>
        <h1>Welcome to the quiz</h1>
        <button>Start new quiz</button>
    </>
}

function QuizPage({quizFactory}: {quizFactory(): Quiz}) {
    const quiz = quizFactory();
    const {question, alternatives} = quiz;
    return <>
        <h1>Quiz</h1>
        <div>Question: {question}</div>
        {alternatives.map(a =>
            <button>{a}</button>
        )}
    </>
}

export function App() {
    const quiz: Quiz = {
        question: "Are you happy today?",
        alternatives: ["yes", "no", "maybe"]
    }
    return <BrowserRouter>
        <Switch>
            <Route path="/quiz">
                <QuizPage quizFactory={() => quiz} />
            </Route>
            <Route exact path="/">
                <HomePage />
            </Route>
            <Route>
                Not found
            </Route>
        </Switch>
    </BrowserRouter>;
}