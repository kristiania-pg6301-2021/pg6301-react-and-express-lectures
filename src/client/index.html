<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <title>Hello Kristiania</title>
  </head>
  <body>
    <div id="app"></div>
  </body>
  <script src="index.jsx"></script>
</html>
