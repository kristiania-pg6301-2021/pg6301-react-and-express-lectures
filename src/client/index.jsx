import * as React from "react";
import * as ReactDOM from "react-dom";
import { Application } from "./Application";

ReactDOM.render(<Application />, document.getElementById("app"));
