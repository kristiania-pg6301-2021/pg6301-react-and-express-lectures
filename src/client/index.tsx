import * as React from "react";
import * as ReactDOM from "react-dom";

import {BrowserRouter} from "react-router-dom";
import {Route, Switch} from "react-router";

interface Quiz {
    question: string;
    alternatives: string[];
}


function HomePage() {
    return <>
        <h1>Welcome to the quiz</h1>
        <button>Start new quiz</button>
    </>
}

function QuizPage({quizFactory}: {quizFactory(): Quiz}) {
    const quiz = quizFactory();
    const {question, alternatives} = quiz;
    return <>
        <h1>Quiz</h1>
        <div>Question: {question}</div>
        {alternatives.map(a =>
            <button>{a}</button>
        )}
    </>
}

function App() {
    const quiz: Quiz = {
        question: "Are you happy?",
        alternatives: ["yes", "no", "maybe"]
    }
    return <BrowserRouter>
        <Switch>
            <Route path="/quiz">
                <QuizPage quizFactory={() => quiz} />
            </Route>
            <Route exact path="/">
                <HomePage />
            </Route>
            <Route>
                Not found
            </Route>
        </Switch>
    </BrowserRouter>;
}


ReactDOM.render(<App />, document.getElementById("app"));