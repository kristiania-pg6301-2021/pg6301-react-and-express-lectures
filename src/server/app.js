/* eslint-disable */
require("dotenv").config();
const path = require("path");
const express = require("express");
const bodyParser = require("body-parser");
const session = require("express-session");
const passport = require("passport");
const GoogleStrategy = require("passport-google-oauth20").Strategy;

function createApp() {
  const app = express();

  app.use(bodyParser.json());
  app.use(
    session({
      secret: "a secret used to encrypt the session cookies",
      resave: true,
      saveUninitialized: false,
    })
  );

  passport.use(
    new GoogleStrategy(
      {
        clientID: process.env.GOOGLE_CLIENT_ID,
        clientSecret: process.env.GOOGLE_CLIENT_SECRET,
        callbackURL: "https://webapps.kristiania.no:3000/oauth2callback",
      },
      function (accessToken, refreshToken, profile, cb) {
        cb(null, profile.displayName);
      }
    )
  );


  app.use(passport.initialize());
  app.use(passport.session());
  passport.serializeUser((user, done) => {
    done(null, user);
  });
  passport.deserializeUser((id, done) => {
    done(null, id);
  });

  app.get("/api/account", (req, res) => {
    const { user } = req;
    if (!user) {
      return res.status(401).send();
    }
    res.json({ user, account: { balance: 240 } });
  });

  app.get("/login", passport.authenticate("google", { scope: ["profile"] }));

  app.get(
    "/oauth2callback",
    passport.authenticate("google", { failureRedirect: "/login" }),
    function (req, res) {
      // Successful authentication, redirect home.
      res.redirect("/");
    }
  );

  app.use(express.static(path.resolve(__dirname, "..", "..", "dist")));

  app.use((req, res, next) => {
    if (req.method === "GET" && !req.path.startsWith("/api")) {
      res.sendFile(path.resolve(__dirname, "..", "..", "dist", "index.html"));
    } else {
      next();
    }
  });
  return app;
}

module.exports = {
  createApp,
};
