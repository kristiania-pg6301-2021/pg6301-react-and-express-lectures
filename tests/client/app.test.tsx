import * as React from "react";

import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";

import {App} from "../../src/client/app";

let container: Element|undefined;
beforeEach(() => {
    // setup a DOM element as a render target
    container = document.createElement("div");
    document.body.appendChild(container);
});

afterEach(() => {
    // cleanup on exiting
    unmountComponentAtNode(container!);
    container!.remove();
    container = undefined;
});



describe("application", () => {
    
    it("shows application home page", () => {
        act(() => {
            render(<App />, container!);
        });
        expect(container?.innerHTML).toMatchSnapshot();
    });
    
});
