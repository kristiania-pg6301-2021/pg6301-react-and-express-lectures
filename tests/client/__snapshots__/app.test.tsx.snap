// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`application shows application home page 1`] = `"<h1>This is the quiz</h1><button>Start new quiz</button>"`;
